nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+                .write(|w| w.packetptr().bits(packet.buffer.as_ptr() as u32));
+        }
+
+        'cca: loop {
+            // start CCA
+            self.radio
+                .tasks_ccastart
+                .write(|w| w.tasks_ccastart().set_bit());
+
+            loop {
+                if self
+                    .radio
+                    .events_ccaidle
+                    .read()
+                    .events_ccaidle()
+                    .bit_is_set()
+                {
+                    // channel is clear
+                    self.radio.events_ccaidle.reset();
+                    break 'cca;
+                }
+
+                if self
+                    .radio
+                    .events_ccabusy
+                    .read()
+                    .events_ccabusy()
+                    .bit_is_set()
+                {
+                    // channel is busy
+                    self.radio.events_ccaidle.reset();
+                    continue 'cca;
+                }
+            }
+        }
+
+        // due to a shortcut the transmission will start automatically so we just have to wait
+        // until the PHYEND event is raised
+        self.wait_for_event(Event::PhyEnd);
+        dma_end_fence();
+
+        self.radio
+            .shorts
+            .modify(|_, w| w.ccaidle_txen().clear_bit());
+    }
+
+    /// Moves the radio from any state to the DISABLED state
+    fn disable(&mut self) {
+        // See figure 110 in nRF52840-PS
+        loop {
+            if let Variant::Val(state) = self.radio.state.read().state().variant() {
+                match state {
+                    STATE_A::DISABLED => return,
+
+                    STATE_A::RXRU | STATE_A::RXIDLE | STATE_A::TXRU | STATE_A::TXIDLE => {
+                        self.radio
+                            .tasks_disable
+                            .write(|w| w.tasks_disable().set_bit());
+
+                        self.wait_for_state(STATE_A::DISABLED);
+                        return;
+                    }
+
+                    // ramping down
+                    STATE_A::RXDISABLE | STATE_A::TXDISABLE => {
+                        self.wait_for_state(STATE_A::DISABLED);
+                        return;
+                    }
+
+                    // cancel ongoing transfer
+                    STATE_A::RX => {
+                        self.radio.tasks_stop.write(|w| w.tasks_stop().set_bit());
+                        self.wait_for_state(STATE_A::RXIDLE);
+                    }
+                    STATE_A::TX => {
+                        self.radio.tasks_stop.write(|w| w.tasks_stop().set_bit());
+                        self.wait_for_state(STATE_A::TXIDLE);
+                    }
+                }
+            } else {
+                // STATE register is in an invalid state
+                unreachable!()
+            }
+        }
+    }
+
+    /// Moves the radio from any state to the RXEN state
+    fn enable_rx(&mut self) {
```

Comment:
There is a possibility that with RADIO SHORTS enabled this function would work in an unexpected way. I would add at least assertion here to verify that selected shorts are disabled.